[Tensorpipe/RPC] tensorpipe RPC agent

jiayisuse opened the pull request:
Implement the initial version of TensorPipe RPC agent, and register to RPC registry to expose to Python interface. As a starter, it utilizes all available TensorPipe transports (shm, uv) and channels (basic, cma).

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+      0, "cma", tensorpipe::channel::cma::CmaChannelFactory::create());
+
+  std::vector<std::string> addresses = {createUniqueShmAddr(),
+                                        "tcp://127.0.0.1"};
```
Ideally tensorpipe could provide a helper to get IP address for given device interface or host names, or return the IP address of the default host name. More specifically like Gloo https://github.com/facebookincubator/gloo/blob/master/gloo/transport/tcp/device.cc#L142. As this is not supported in tensorpipe right now, we could have a couple of work around:
* Retrieve the default IP address on rpc agent layer
* Let user specify IP address by rpcBackendOption
* Hard code 127.0.0.1 for now for single machine only

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+
+        // Allocate memory and fill in pointers
+        Message rpcMessage = tensorpipeAllocateMessage(tpMessage);
+        TORCH_INTERNAL_ASSERT(
```
Will this error propagate back to the sender?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+        for (int i = 0; i < rpcMessage.tensors().size(); i++) {
+          auto& rpcTensor = rpcMessage.tensors()[i];
+          auto& tpTensor = tpMessage.tensors[i];
+          tpTensor.data = (uint8_t*)(rpcTensor.data_ptr());
```
In `wireSerialization` in `rpc/utils.cpp`, we would send out the entire storage if the the tensor data is a significant portion of the storage. So, there is chance that this tensor might not be contiguous (@jjlilley please correct me if I am wrong). Do we need to check that here?

jjlilley replied:
Yes, though in the ``tensorpipeSerialize()`` function rather than here.

The code should probably rpc::cloneSparseTensors() here to avoid that situation... with training data, it's common to send 1 out of 100 rows in an rpc. Locally, it's implemented just as a view into the original data with a stride, but you want to re-copy to avoid 100x blowup in size.

lw replied:
I'd like to understand better what the RPC API contract is here. If the agent is allowed to clone sparse tensors and "compress" the storage by only sending over the parts that are "visible", this means that the receiver will get a slightly different tensor than the one that was sent, namely its strides may be different and its `.storage()` may have a different size. Is this intended/expected?

Is there a generic helper function that can be shared by all agents to make sure this conversion is done correctly and consistently? It looks like it becomes complicated really quickly. Think, for example, what would happen if one sends a self-overlapping tensor (e.g., a 2x3 tensor, backed by a storage of size 5, with strides (2, 1), which means that [0,2] is the same location as [1,0]).

jjlilley replied:
@lw That's a good point to raise.

- The existing logic was effectively added in December when we realized that we were sending 100x too much data in the wire in the testing training; we were completely saturating the NICs, trainer qps was quite low, and it was quite clear that we had to do address this. It's apparently a common pattern to do this for training samples (have a large tensor of 100 training rows, but sample/send one at a time). 

- Since then, the team has realistically been sufficiently occupied with various training performance issues that rationalizing the api contract honestly hasn't been high pri.

- So, presently, the other rpc_agents do not guarantee that the resulting strides and underlying .storage() are identical. IMO, it would be reasonable to add an default-off option to do so, if there's a need for it. But e.g. the platform in general doesn't even guarantee that .storage() even exists or is usable.

mrshenli replied:
This looks like an important thing we need to clarify in our API doc. Let me create an issue to track this.

mrshenli replied:
#35766

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+  if (!isRunning_) {
+    return;
+  }
+
```
Looks like this function can run on a separate thread? If so, will it be a problem if `isRunning_ ` is set to False (shutdown) after the above check?

jjlilley commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,89 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    TORCH_INTERNAL_ASSERT(
+        tensorDataVec.size() == 1, "There should be single pickled tensor");
+    const auto& tensorData = tensorDataVec.front();
+    // NOTE: tensorData shares the same storage with rpcMessage tensor
```
1. How is the code guaranteeing the lifetime of the source Tensor?
   I don't see a place in tpMessage where you're persisting the tensors.

2. This copy avoidance on write isn't universally safe, since it's legal to call with torch::from_blob() tensors, where the user can [unwisely] choose to manually manage tensor memory. I fixed a related bug in this space in ProcessGroupAgent earlier this week, probably won't be the last.

To mitigate, ThriftRpcAgent's WireSerializer checks whether the tensor has a deleter (grep for storageHasDeleter()). If there's no deleter, it has to assume that the user is manually managing memory, and has to force a copy.

jiayisuse replied:
@jjlilley Thanks for the insights!
1. From the discussion with @mrshenli, seems users are supposed to keep the tensor buffer lifetime and content consistency.
2. I see. Will do the same check to take sparse tensor and storageHasDeleter() into considerations.

jjlilley replied:
Re: 1- I want to make sure we're on the same page here.

- If the caller calls ``future = agent.send(dest, {tensor})``, and then the tensor goes out of scope, how is TensorPipes agents keeping the tensor alive to get the bits out to the kernel, particularly with the io send on another thread and might suffer arbitrary delays?

- fwiw, both ProcessGroupAgent and ThriftRpcAgent will persist the passed-in Tensor data as long as is necessary to get the bits into the kernel.
   They do it in different ways (and Thrift avoids a copy), but they still keep the Tensor refcount alive across the caller->io_thread transition.
   Neither guarantees what bits will be sent if the caller mutates the tensor immediately after the async call starts, but they do guarantee no read-after-free.

It's possible that I missed the mechanism, but it's going to be fairly essential, even in a minimalist impl, to maintain the lifetime of the shared buffer appropriately.

(The (2) point is fairly minor in comparison)

jjlilley commented:
Aside: one nice feature of TensorPipe vs ThriftRpcAgent, when TensorPipes eventually makes it to prod - for large Tensors, Thrift has to read Tensors in 16MB chunks and re-copy at the end, due to its use of the rsocket protocol. (Rsocket header only has 3 bytes for the length field, and hence has to fragment when data is bigger than that).

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
+  for (const std::pair<std::string, worker_id_t>& kv : opts_.workerNameToId) {
```
@mrshenli In the `init_rpc` function (see [docs here](https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.init_rpc)) the rank seems to be optional (since it takes -1 as default), although all examples do actually pass in a rank. So can RPC agents assume that a valid rank will always be provided, or do they need to be able to handle the case of each node just knowing its name (and the world size)?

mrshenli replied:
Yes, we can expect all callers to pass in a valid rank to `init_rpc` for now. But it is unclear to me whether we can expect the same thing when support dynamic worker join/leaves after enable elasticity. We probably can let the store to assign rank so sth then.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+    onListenerAccepted(error, pipe);
+  });
+
+  // Arm for server read
```
"Arming" as in "preparing it to fire", where "it" is the callback invocation. Callbacks are one-shot in TensorPipe, so each time one triggers it needs to be reloaded.

mrshenli replied:
Ahh! "preparing it to fire" that's a good and interesting term to use here.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,367 @@
+
+  for (const auto& p : workerNameToInfo_) {
+    const auto& name = p.first;
+    auto nodeAddrData = addressStore_->get(name);
```
@agolynski is the goal to modify existing c10d Store for elasticity? If yes, this can ben the first step towards that?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+#include <tensorpipe/core/context.h>
+#include <tensorpipe/core/listener.h>
+#include <tensorpipe/core/pipe.h>
+#include <torch/csrc/distributed/rpc/python_rpc_handler.h>
```
Doesn't seem we need this header. This problem exists in `process_group_agent.h` as well. Create #35767 to remove that.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = std::to_string(static_cast<int>(rpcMessage.type())) +
```
c10::str here as well?

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  for (at::Tensor tensor : tensors) {
```
use reference here? Tensor internally contains a `intrusive_ptr` instance.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    // Keep original user tensors and cloned sparse tensors
+    reservedTensors.push_back(tensor);
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler(
```
Question, is there any reason for doing this for every individual tensor instead of using one pickler for the entirer tensors list? Is it because we cannot tell the boundary between tensor metas in the writer lambda?

jiayisuse replied:
Iterating the tensors is mainly because of current tensorpipe::Message interface. We need to fill metadata separately to tensorpipe::Message::tensors according to Message interface. Or we can hack by putting all tensors' metadata into the first tensor metadata. As the first version of implementation, we may would like to keep the design clean.
It gets explained with more details and future interface change plan in the design quip doc, section Data Conversion

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    const auto& tensorData = tensorDataVec.front();
+    // Enforce memory copy if tensor is created from torch::from_blob, means
+    // that the tensor doesn't own the memory.
+    if (!tensorData.storageHasDeleter()) {
```
Would I be correct if I assume we don't really need the `WriteableTensorData` to check the storage deleter? If so, going through the full pickler seems like an overkill, and the `getWriteableTensorData` seems always copy CUDA tensors to CPU, is this intentional for TensorPipe? 

This is certainly OK for now, I wonder if we should extract out the meta parsing logic from the pickler into a utility function.

jiayisuse replied:
Regarding to CUDA -> CPU copy, yes. TensorPipe doesn't support CUDA channel yet and need to move to CPU always. Once CUDA copy is enabled in TensorPipe, we would need to either modify pickler or create a separate util to keep CUDA tensors on GPU.
letting tensors go through pickler is mainly to serialize metadata.. And correct me if I'm wrong, pickler shouldn't do memory copy of tensor storage (for CPU tensors)?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+  void addGilWaitTime(const std::chrono::microseconds gilWaitTime) override {}
+
+ private:
+  const std::string& findWorkerURL(const WorkerInfo& worker);
```
this can be const fun?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+  std::unordered_map<std::string, std::string> workerNameToURL_;
+
+  std::shared_ptr<::c10d::Store> addressStore_;
+  TensorPipeRpcBackendOptions opts_;
```
These can be const for now?

mrshenli reviewed (commented):
Code looks good to me. Thanks for the huge effort! There are some more comments in the internal diff, please resolve those with reviews as well. 

I understand this is not mean to the a comprehensive PR that ships everything, but still would like to understand how we plan to land this into OSS. For now one big missing part is testing from Python API. (cc @osalpekar for rpc-backend-generic testing infra) Do we plan to add any Python tests in this PR, or is that future work?

mrshenli reviewed (changes requested):
Build error is real.

osalpekar commented:
> Code looks good to me. Thanks for the huge effort! There are some more comments in the internal diff, please resolve those with reviews as well.
> 
> I understand this is not mean to the a comprehensive PR that ships everything, but still would like to understand how we plan to land this into OSS. For now one big missing part is testing from Python API. (cc @osalpekar for rpc-backend-generic testing infra) Do we plan to add any Python tests in this PR, or is that future work?

@mrshenli I'll add the `rpc_tests` and `dist_autograd_tests` with the Tensorpipe backend in a follow-up PR!

lw commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+          tpTensor.metadata.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
```
Following D20749423, this `nullptr` should be removed.

jiayisuse replied:
Yes, it's removed from the 1st diff in stack.. Don't know why it's not updated here

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+  // https://github.com/pytorch/pytorch/issues/27647
+  void join() override;
+  void sync() override;
+  void start() override;
```
After D20074714, this should now be called `startImpl`.

lw reviewed (commented):
Left comments on a couple of issues I encountered when rebasing this on master.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,380 @@
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage_[requestMessage.id()] = futureResponseMessage;
+
+  pipeWrite(
```
I think the lock should be released before calling `pipeWrite`, because the callback be reentrant (i.e., be called inline), and it will try to acquire the lock, which might deadlock.

jiayisuse replied:
We discussed in out 1:1 a bit. Successful messages would invoke callbacks from looping threads. Only error callbacks could be invoked inline. In this integration, we are not doing any retry on send() from error callback. So we shouldn't have deadlock?

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,380 @@
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
```
The pipe received in this callback is not stored on a member field of the agent, it's just passed around by reference and sometimes captured by copy in a callback. Would it perhaps be more robust to store it, to make sure it's kept alive no matter what?

jiayisuse replied:
I see your point. The pipe would be kept alive in the next respond() calls. It will be gone when gets an error say closed by the peer. If we would like to keep a copy in agent, it may be overhead to handle the destruction if the connection is closed, or just keep it alive until agent is destructed?

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,380 @@
+  shutdown();
+}
+
+void TensorPipeAgent::shutdown() {
```
After https://github.com/pytorch/pytorch/pull/36239, this should now be `shutdownImpl`. The destructor and `join` can still call `shutdown`

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,380 @@
+#include <caffe2/torch/csrc/distributed/rpc/tensorpipe_agent.h>
```
The path should be relative to `torch`, not `caffe2` for OSS code. This should also help fix the clang-tidy error!

jiayisuse commented:
Adding tensorpipe as submodule: https://github.com/pytorch/pytorch/pull/36636

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+                std::lock_guard<std::mutex> lock(mutex_);
+                // One read error will lead all following callbacks to be
+                // invoked with error, and shouldn't read here.
+                TORCH_INTERNAL_ASSERT(!readError_, "Shouldn't in error state");
```
I suspect this assert may fail. If one pipe gets an error, the global `readError_` will be set, but other pipes may still continue to work properly. I think that when we receive an error on a pipe we should only fail the messages being sent on that pipe, and keep the other ones untouched. This is the most robust solution, especially for elastic training scenarios where machines are expected to come and go. Alternatively, to keep it simple at first, we can proactively close all pipes when one of them returns an error.

pritamdamania87 commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+}
+
+std::string TensorPipeAgent::createUniqueShmAddr() {
+  return c10::str(
+      "shm://tensorpipe_rpc_agent_",
+      std::this_thread::get_id(),
+      "_",
+      nextShmAddressID_++);
```
It's possible that we generate the same shm addr for two different processes on the same host since the first time `nextShmAddressID_` would be 0 for both and since thread ids can be recycled, `std::this_thread::get_id()` can potentially return the same id. Also, I'm not sure if `std::this_thread::get_id()` is guaranteed to be unique across different processes.

I think we should probably just refactor and reuse: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_context.h#L61 and get rid of `std::this_thread::get_id()` and `nextShmAddressID_` here.

jiayisuse replied:
Seems it's using getWorkerId() and nextLocalId to create globally unique identifier. If we have multiple training groups on a single host, worker_id could conflict as well? LocalId also doesn't provide uniqueness across multiple processes. Maybe we can use process ID + thread ID + local ID? As shm is just for linux for now, it should be fine to have some platform specific code?

pritamdamania87 replied:
Worker id is supposed to be unique across a training group. And I think different training groups should use different store instances or if they use the same store instance, they should use a different prefix in the store to avoid conflicts.

pritamdamania87 commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+  std::map<std::string, worker_id_t> workerNameToId;
+};
+
+class TensorPipeAgent : public RpcAgent {
```
Add some docs to this class.

pritamdamania87 commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,111 @@
+
+  const std::shared_ptr<::c10d::Store> addressStore_;
+  const TensorPipeRpcBackendOptions opts_;
+  std::atomic<bool> isRunning_;
```
Can we reuse `rpcAgentRunning_` from RpcAgent instead of having another flag here? cc @osalpekar

osalpekar replied:
Yes, we can just use `rpcAgentRunning_` wherever needed. `start` and `shutdown` in `RpcAgent` take care of setting and unsetting this flag.

pritamdamania87 commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+    : RpcAgent(
+          WorkerInfo(std::move(selfName), std::move(selfId)),
+          std::make_unique<RequestCallbackImpl>(),
+          opts.rpcTimeout),
```
I'm guessing we don't honor the rpcTimeout in TensorPipe yet? Can we add a TODO for this?

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -39,6 +40,18 @@ TORCH_API std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  std::vector<std::vector<uint8_t>> copiedTensors;
+};
+
+TORCH_API TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage);
+
+TORCH_API Message
+tensorpipeAllocateMessage(const tensorpipe::Message& tpMessage);
```
Add appropriate docs for public methods.

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  c10::List<at::Tensor> tensors = cloneSparseTensors(rpcMessage.tensors());
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
```
We should add a note to `tensorpipeSerialize` docs that the caller needs to keep rpcMessage alive for the returned TensorPipeEntry to be valid since TensorPipeEntry has references to the original rpcMessage.

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -39,6 +40,18 @@ TORCH_API std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
     const void* data,
     size_t data_size);
 
+// Tensorpipe message conversion
+struct TensorPipeEntry {
+  tensorpipe::Message message;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
+};
+
+TORCH_API TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage);
+
+TORCH_API Message
```
Instead of putting this in utils, might be better to have something like torch/csrc/distributed/rpc/tensorpipe/serialize.{h, cpp}. The rest of the tensorpipe code can also go inside the tensorpipe/ directory.

jiayisuse replied:
I'm not sure if it deserves a separate folder for tensorpipe. As these serialization and deserialization functions doing similar with wireSerialize, maybe let's keep in this file for now

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -39,6 +40,18 @@ TORCH_API std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+// Tensorpipe message conversion
+struct TensorPipeEntry {
+  tensorpipe::Message message;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
```
Would be nice to add docs about what reservedTensors and copiedTensors are.

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    // Enforce memory copy if tensor is created from torch::from_blob, means
+    // that the tensor doesn't own the memory.
+    if (!tensorData.storageHasDeleter()) {
+      std::vector<uint8_t> storageData(tensorData.sizeInBytes());
+      memcpy(storageData.data(), tensorData.data(), storageData.size());
+      tpTensor.data = storageData.data();
+      copiedTensors.push_back(std::move(storageData));
+    } else {
+      tpTensor.data = (uint8_t*)(tensorData.data());
+    }
+    tpTensor.length = tensorData.sizeInBytes();
+    tpMessage.tensors.push_back(std::move(tpTensor));
```
Would this work for tensors that are non-contiguous?

jjlilley replied:
None of the RpcAgents support sparse/non-contiguous tensors. I believe it was decided late last year to defer that work to the Pickler, which would make it work in many more situations as well.

My impression is that it's not currently high-pri in Pickler, though probably doable if there's a motivating use case.

My suggestion would be to get v1.0 in (and e.g. building on OSS/github) before we worry about cases like this.

pritamdamania87 replied:
I thought the  pickler does handle non-contiguous (dense non-contiguous tensors not referring to sparse tensors) tensors by storing the strides appropriately: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/pickler.cpp#L373. I did add a simple test as well to check that this seems to work for our current RPC agents: https://github.com/pytorch/pytorch/pull/36705 (and since this logic is similar, it probably would work here as well?)

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = c10::str(
+      static_cast<int>(rpcMessage.type()),
+      ":",
+      static_cast<int>(rpcMessage.id()));
```
Can we avoid doing this since type() and id() have fixed size? Just make the first 4 bytes of metadata type() and next 4 bytes id()?

jjlilley replied:
The code here is wrong, since it truncates off the top 32 bits off the rpc::Message's int64_t.

I agree with Pritam that the string parsing might be a little heavyweight. (If the rest of the message were text, I'd lean towards a text encoding, but here, it seems like binary might be nice)

The main caveats are:
 - make sure to leave 8 bytes for the id()
 -  don't know if we care about byte-order here? it might be good form to htonl/etc if you want, though basically everybody uses arm/x86 these days :)

pritamdamania87 replied:
Regarding byte-order I used `THP_encodeInt64Buffer` here https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/rpc_messages/rpc_with_autograd.cpp#L80, we can do something similar for this.

pritamdamania87 replied:
> The code here is wrong, since it truncates off the top 32 bits off the rpc::Message's int64_t.

Good catch!

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+
+Message tensorpipeAllocateMessage(const tensorpipe::Message& tpMessage) {
+  // Payload, message type and message id
+  std::vector<char> payload(tpMessage.length);
+  size_t delimiterPos = tpMessage.metadata.find(":");
+  TORCH_INTERNAL_ASSERT(
+      delimiterPos != std::string::npos,
+      "message metadata must be delimitted by \":\" `",
+      tpMessage.metadata,
+      "`");
+  int mtypei = std::stoi(tpMessage.metadata.substr(0, delimiterPos));
+  uint64_t mId = std::stoull(tpMessage.metadata.substr(delimiterPos + 1));
+  MessageType mtype = static_cast<MessageType>(mtypei);
```
We can avoid all this messy string ops if we make the first 4 bytes the message type and next 4 bytes the message id.

pritamdamania87 commented:
Also, looks like the OSS CI builds are broken.

jiayisuse commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -348,6 +349,48 @@ If the future completes with an error, an exception is thrown.
+          py::arg("name"),
+          py::arg("address_store"),
+          py::arg("tensorpipe_backend_options"))
+      .def(
+          "join",
+          &TensorPipeAgent::join,
+          py::call_guard<py::gil_scoped_release>())
```
Process Group agent is exporting. Following it to do the same. Shall we export?

pritamdamania87 replied:
Actually, we probably need to export this for now since we call `join` in rpc.shutdown.

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+void TensorPipeAgent::onListenerAccepted(
+    const tensorpipe::Error& error,
+    std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  // TODO: Handle listener error
+  if (error) {
+    LOG(WARNING) << "got error from listener: " << error.what();
+    return;
+  }
```
Right now it would be just timeout. We will handle these errors in following diffs

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+        }
+
+        // Arm for next read
+        respond(pipe);
```
This is for a given pipe to respond multiple requests from the client. Accept is to create a new pipe to serve a new client.

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+
+        // Shortcut if immediately done
+        if (futureResponseMessage->completed()) {
+          sendCompletedResponseMessage(
```
For now there is no exception

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+            pipe,
+            [this](const tensorpipe::Error& error, Message&& responseMessage) {
+              if (error) {
+                LOG(WARNING) << "Read response error: " << error.what();
```
Do you mean hanged or terminated? Normally peer would get EOF error upon remote node exits. If remote node somehow stops responding, all the requests to that node (via the pipe) won't be served, and this piece of code won't be called. I believe that's where timeout should chime in.

pritamdamania87 replied:
I meant if the remote node was terminated. In what case would we see this error then? Would it be when we receive an error reading from the socket?

pritamdamania87 commented:
@jiayisuse As Jeremy pointed out, we can try to land v1.0 of this soon and iterate on top of it. So feel free to address some of the comments in follow up PRs (if they're not critical correctness issues).

mrshenli commented:
looks like build will need #36636 to land first

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,401 @@
+  });
+
+  // Arm for server read
+  respond(pipe);
```
We can see TensorPipe as single loop thread. Think of a socket. It's designed to create multiple pipes (pipe pool) to achieve better performance. In current solution, say we have trainer:0 sending rpc requests to es:0:0-9, there would be 10 pipes. To bump a single pair of trainer:0 to es:0:0, we can have a pool of pipes. I'll add a TODO for this

pritamdamania87 replied:
Another option is to have the single loop thread only accept connections and write responses back to the socket. For deserialization of request, calling RequestCallbackImpl and serialization of response, is it possible to defer this work to a separate threadpool?

jjlilley replied:
> Another option is to have the single loop thread only accept connections and write responses back to the socket. For deserialization of request, calling RequestCallbackImpl and serialization of response, is it possible to defer this work to a separate threadpool?

I suspect that such an approach would be trickier to use here, since the tensorpipes approach attempts to read directly into the result tensor buffer - serialization is more tied with IO than e.g. in the Thrift approach (but OTOH tensorpipe deserialization is somewhat lighter weight)

If we need more bandwidth between a pair of endpoints than a single thread allows, ideally we might have a parallel option to Thrift's PooledRequestChannel here, where a connection pair can effectively "trunk" across a set of threads/sockets (and avoid some head-of-line-blocking issues, though giving up fifo/ordering)

jiayisuse commented:
@pritamdamania87 @mrshenli Regarding to the build issue, yes, it will need https://github.com/pytorch/pytorch/pull/36636 to land, plus tensorpipe building integration with pythroch

beauby commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,6 +353,109 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  tpMessage.metadata.resize(2 * sizeof(int64_t));
+  int64_t mtype = static_cast<int>(rpcMessage.type());
+  int64_t mId = rpcMessage.id();
+  memcpy(tpMessage.metadata.data(), &mtype, sizeof(int64_t));
+  memcpy(tpMessage.metadata.data() + sizeof(int64_t), &mId, sizeof(int64_t));
```
The `data()` method of `std::string` returns a `const char*`. Would `tpMessage.metadata.append(reinterpret_cast<unsigned char*>(&mtype), sizeof(mtype));` work instead?
Also, nit: inconsistent capitalization (`mtype` vs `mId`).

pritamdamania87 commented:
@jiayisuse Is this ready for another round of review? You can click the "Re-request review" next to a reviewer to request another review.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <torch/csrc/distributed/rpc/python_rpc_handler.h>
 #include <torch/csrc/distributed/rpc/rpc_agent.h>
 #include <torch/csrc/distributed/rpc/rref_context.h>
+#include <torch/csrc/distributed/rpc/tensorpipe_agent.h>
```
Not requesting for change in this PR, we probably should guard this and the `process_group_agent.h`  above (and other related parts) with `ifdef`. Added #37782 to track.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -444,6 +445,48 @@ If the future completes with an error, an exception is thrown.
           py::call_guard<py::gil_scoped_release>());
 
+  // Base class: torch.distributed.rpc.RpcBackendOptions.
+  py::class_<TensorPipeRpcBackendOptions>(
```
(Not request changes in this PR). We will need API docs for this type. Added #37783 to track.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -444,6 +445,48 @@ If the future completes with an error, an exception is thrown.
+      module, "TensorPipeRpcBackendOptions", rpcBackendOptions)
+      .def(py::init<>())
+      .def_readwrite(
+          "worker_name_to_id", &TensorPipeRpcBackendOptions::workerNameToId);
```
Added #37784 to track. This is partly related to the elasticity project, as both might require some notion of a store to share the worker information. cc @agolynski

jiayisuse replied:
Yes, currently we will need users to provide this name-id mapping through rpc backend options. Here you mean we automatically collect these information through c10::store, remove, add or update nodes?

mrshenli replied:
it doesn't have to be `c10d::store`, it's just one option we discussed before, where the `TensorPipeRpcBackendOptions ` can contain some `init_method` to initialize/find the store, and then every worker writes its own name->id into the store. This might require some changes to the existing `c10d::Store`. If TensorPipe has better options to support this feature, it's definitely welcome.

I am raising this concern because it seems asking applications to provide a name->id map might be a regression on API convenience.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,127 @@
+#endif
+
+  // TensorPipe read function that could be used to read response messages
+  // by client, and read requests request messages by server.
```
remove "requests"?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -0,0 +1,127 @@
+
+  // We need map one keyed on name and one on id for easy lookup.
+  std::unordered_map<worker_id_t, WorkerInfo> workerIdToInfo_;
+  std::unordered_map<std::string, WorkerInfo> workerNameToInfo_;
```
Given that the options already has a `workerNameToId` field, I assume keeping two maps here is to avoid an additional lookup?

jiayisuse replied:
Yes, for easier lookup to serve override functions getWorkerInfo(const std::string& workerName) and getWorkerInfo(worker_id_t workerId)

mrshenli commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -39,6 +40,29 @@ TORCH_API std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+// The passed-in tensorpipe message is partial, which just contains
+// necessary information for memory allocation, like payload length
+// and tensor metadata. The returned RPC message doesn't have any
+// data, but would be valid after tensorpipe finishs data stransfer.
```
stransfer -> transfer?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,411 @@
+      context_(std::make_shared<tensorpipe::Context>()),
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
```
Add a `TORCH_CHECK` to verify that `selfName` is indeed in the map?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,411 @@
+
+  const auto& url = findWorkerURL(toWorkerInfo);
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // See if we already have a connection to this address or not
+  auto it = connectedPipes_.find(toWorkerInfo.id_);
+  if (it == connectedPipes_.end()) {
+    std::tie(it, std::ignore) = connectedPipes_.emplace(
+        toWorkerInfo.id_, ClientPipe(context_->connect(url)));
+  }
+  ClientPipe& clientPipe = it->second;
+  auto& pendingResponseMessage = clientPipe.pendingResponseMessage_;
+
+  std::shared_ptr<FutureMessage> futureResponseMessage =
+      std::make_shared<FutureMessage>();
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage[requestMessage.id()] = futureResponseMessage;
+
+  // Don't need to hold lock while calling tensorpipe API.
+  lock.unlock();
```
Use a `lock_guard` here?

mrshenli reviewed (approved):
This is awesome! Thanks a lot for putting this together. 

I just add some trackers for follow up work items, and some cosmetic comments. 

Let's wait for all OSS tests to pass before landing.

mrshenli reviewed (changes requested):
One minor revision request: shall we remove internal-only info from the PR description? We need to do this for both PR summary and phab diff summary, as the former is where we track OSS discussion and the latter is what goes into the git commit message. 

Sorry about the churn.

jiayisuse commented:
Thanks for the review, guys! Will wait for the CI to finish before addressing the remaining comments. CI takes some time to run to the end

lw commented:
Since the only blocker to get this landed is to change the summary, I guess we can just fix that on Phabricator and land the diff, without re-exporting it and waiting again for the CI to complete, since nothing should change in the CI signal, except for flaky tests. Right?

mrshenli commented:
> Since the only blocker to get this landed is to change the summary, I guess we can just fix that on Phabricator and land the diff, without re-exporting it and waiting again for the CI to complete, since nothing should change in the CI signal, except for flaky tests. Right?

Right, we don't need to export this again.

mrshenli commented:
All tests pass, this is good to go, let's land!

jiayisuse committed:
[TensorPipe/RPC] Serialize and deserialize message

Differential Revision: D20084125

fbshipit-source-id: 67a0ce2237e9f82415f604a94a28ed01ac193076

jiayisuse committed:
[Tensorpipe/RPC] tensorpipe RPC agent (#35483)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/35483

Implement the initial version of TensorPipe RPC agent, and register to RPC registry to expose to Python interface. As a starter, it utilizes all available TensorPipe transports (shm, uv) and channels (basic, cma).

Test Plan:
https://our.intern.facebook.com/intern/diffusion/FBS/browse/master/fbcode/experimental/jiayisuse/tensorpipe_rpc
  export MASTER_ADDR=127.0.0.1
  export MASTER_PORT=28500
  buck build mode/dev-nosan mode/no-gpu //experimental/jiayisuse/tensorpipe_rpc:main
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/main.par
  buck build mode/dev-nosan mode/no-gpu //experimental/jiayisuse/tensorpipe_rpc:benchmark
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/benchmark.par

Multiple connections with async echo
  ./buck-out/gen/experimental/jiayisuse/tensorpipe_rpc/async_echo.par

Reviewed By: lw

Differential Revision: D20088366

fbshipit-source-id: 97dde666bef9288689b9c68125f4443d612288db

rohan-varma commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,411 @@
+                clientPipe.pendingResponseMessage_.erase(it);
+              }
+
+              threadPool_.run(
```
Maybe a noob q, but why is marking the future as complete/error deferred to a thread pool task? I don't think we're doing any I/O/async/heavy work here, so maybe it's best to just run this in line?

jjlilley replied:
Marking a future as complete can entail an unbounded amount of work/processing that doesn't need to be in an IO thread.

lw replied:
Yes, and in fact in #39010 we make sure that all futures are completed from the thread pool (in this PR we had missed a couple).
